### PR TITLE
Updating CTL version, fixing for breaking changes

### LINKED ISSUE #37: Update CTL and fix for breaking changes
In particular, [this PR](https://github.com/Plutonomicon/cardano-transaction-lib/pull/731) was recently merged, so if we want to keep using updates from CTL we'll have to update for those breaking changes first.

### PR DESCRIPTION
**Pull request by rynoV**
Closes #37 and #30 

Changes:

- Use updated CIP25 metadata
- Use updated contract initialization https://github.com/Plutonomicon/cardano-transaction-lib/pull/731
- Stop using `Contract`'s `extraConfig`
- Update `getWalletBalance` so it doesn't cause a major error, see https://github.com/mlabs-haskell/seabug-contracts/pull/39/commits/e04e2f2d8028bc0dcb11eb3dbd3c3b03527e79da and https://github.com/mlabs-haskell/nft-marketplace/tree/calum/contracts-updates

TODO:

- [x] Move away from `extraConfig`
  - This feature of CTL is [going to be removed eventually](https://github.com/Plutonomicon/cardano-transaction-lib/issues/734), I think I may as well refactor while I'm doing this migration
  - I tried a custom monad stack [here](https://github.com/mlabs-haskell/seabug-contracts/commit/6c7023b1f1f8b19f64c74e0bb9f655bdbd729438), it's not much better than just passing an extra parameter to the contracts that need the `projectId`
    - We could make it nicer by implementing our own `liftX` functions to make interoperating with all of CTL's `Contract` functions easier, but I don't think it's worth it right now
- [x] Make sure to properly reuse contract environment (websocket connections, etc.)
  - This PR is on par with master in terms of the number of websocket connections
  - It might be possible to refactor so that the contract environment is reused across contract calls, but I think that's premature optimization at this point
- [ ] Hopefully integrate fixes for https://github.com/Plutonomicon/cardano-transaction-lib/issues/803 because we're also seeing the issue described there
  - The issue isn't breaking the app though, so we could merge without it

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -36,9 +36,9 @@ exports.callMarketPlaceFetchNft = async (config, args) => {
 /**
  * Returns a promise containing the connected wallet's balance.
  */
-exports.getWalletBalance = async () => {
+exports.getWalletBalance = async (config) => {
   const sb = await seabug;
-  return sb.getWalletBalance();
+  return sb.callGetWalletBalance(config);
```
Comment (samuelWilliams99) [inline index.js:41]: might this fix #36 ?

File: src/Seabug/CallContract.purs [hunk 2]
```diff
@@ -71,12 +68,20 @@ import Serialization.Hash
   )
 import Types.BigNum as BigNum
 import Types.Natural as Nat
-import Wallet (mkNamiWalletAff)
+
+liftBuilder :: forall a. Either Error a -> Aff a
+liftBuilder = liftEffect <<< liftEither
```
Comment (samuelWilliams99) [inline src/Seabug/CallContract.purs:73]: You may be able to just use `liftEither` by itself here, theres a `MonadThrow Error Aff` instance

File: src/Seabug/Contract/CnftMint.purs [hunk 3]
```diff
@@ -69,13 +69,16 @@ mintCnft (MintCnftParams params) = do
...
     [ Cip25MetadataEntry
         { policyId: policyHash
-        , assetName: tn
-        , imageUris: NonEmpty.singleton params.imageUri
+        , assetName: wrap tn
+        , name
+        , image: params.imageUri
         , mediaType: Nothing
-        , description: [ params.name, params.description ]
+        , description: Just params.description
```
Comment (samuelWilliams99) [inline src/Seabug/Contract/CnftMint.purs:81]: This seems like an API change? will the frontend handle this correctly?

**Review by samuelWilliams99 — COMMENTED** [review]
I assume merging this will mean we'll need to update the ctl ref in the monorepo as well?

File: index.js (hunk 1 above)
Comment (rynoV) [reply index.js:41]: No it doesn't fix that one, for that one a different function `connectWallet` was removed that we need to add back

Comment (samuelWilliams99) [reply index.js:41]: gotcha

**Comment by rynoV** [comment]
> I assume merging this will mean we'll need to update the ctl ref in the monorepo as well?

It would probably be for the best. Maybe we should merge into some kind of staging branch instead?

**Comment by samuelWilliams99** [comment]
Perhaps, but it's a relatively uncommon change, I think it's fine to just pr both of them
the subrepo should be commit locked anyway

File: src/Seabug/Contract/CnftMint.purs (hunk 3 above)
Comment (rynoV) [reply src/Seabug/Contract/CnftMint.purs:81]: The description and name aren't actually parsed from the metadata, they're retrieved from the database. However the image field is parsed from the metadata, and that format changed from an array to a string, so I updated the parsing for it. According to the cip25 standard the image field can be a string or an array, so I made the parsing handle both possibilities, and this also means the nfts minted before this update are still useable

**Comment by rynoV** [comment]
Tested https://github.com/mlabs-haskell/seabug-contracts/pull/39/commits/866fadd35bd2b237876be1fced23b7dbf91788e7: 

- cnft: https://testnet.cardanoscan.io/transaction/5e91f15855627d1864f8816202a2ed277bcd1c67ab0a263cb44aaa8379d0e61d
- sgnft: https://testnet.cardanoscan.io/transaction/02804e732c6244a564236f536fb5a070a87bbb79b637dc8f109c7892551446aa
- purchase: https://testnet.cardanoscan.io/transaction/1b1b9919dcfd7a10cde0c8ac2482316cf6a13405748f7589f3817cd3d00b1fd5

File: src/Seabug/Metadata.purs [hunk 4]
```diff
@@ -102,12 +103,21 @@ getIpfsHash
...
+    getNestedAeson r
+      [ "onchain_metadata", currSymKey, tokenNameKey, "image" ]
+  except $ lmap BlockfrostOtherError $ caseAeson
+    ( constAesonCases
+        ( Left "Could not parse image field as string or array"
+        ) # _
+        { caseArray =
+            ( lmap (const "Image array does not contain a string") <<<
+                decodeAeson
+            ) <=< note "Image field contains empty array" <<< head
```
[the remark this answers is not in the document]
Comment (rynoV) [reply src/Seabug/Metadata.purs:116]: For metadata created by us, we shouldn't be getting any images which are long strings split up, unless we change the length of our image value.

We could use CTL's cip25 modules to parse this and that should handle long strings being split into arrays, but then the NFTs minted previously would no longer be valid. I guess it's okay though, the old nfts are just for testing anyways

[the remark this answers is not in the document]
Comment (samuelWilliams99) [reply src/Seabug/Metadata.purs:116]: We'll need to support arbitrary cNFTs that adhere to the standard, as we won't always be the ones minting them. In future, artists will want a way to take their existing nft and import it to the protocol.

[the remark this answers is not in the document]
Comment (samuelWilliams99) [reply src/Seabug/Metadata.purs:116]: #42

**Review by samuelWilliams99 — APPROVED** [review]

**Pull request merged**